From 87bac7fb610f81157ae9de719a7d1cec8ed7e2ea Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Thu, 26 May 2016 16:36:58 +0300
Subject: [PATCH 0244/1345] net: mvpp2x: prefetch two rx_descriptors, every
 second packet

commit  fb28400ef897d4f952b626c6dd517dcaa104fd16 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

-rx_descriptors is 32B, so prefetch is done in couples.
-Gives better performance during tested throughput benchmarks.

Change-Id: I7f27664f57ed65b014daeee47de268bdc47959db
Reviewed-on: http://vgitil04.il.marvell.com:8080/30029
Reviewed-by: Yuval Caduri <cyuval@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.h |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.h b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.h
index d02736a..4d2770f 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.h
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_hw.h
@@ -163,7 +163,9 @@ static inline void mv_pp2x_rxq_status_update(struct mv_pp2x_port *port,
 	int rx_desc = rxq->next_desc_to_proc;
 
 	rxq->next_desc_to_proc = MVPP2_QUEUE_NEXT_DESC(rxq, rx_desc);
-	prefetch(rxq->first_desc + rxq->next_desc_to_proc);
+	/* For uneven descriptors, fetch next two descriptors (2*32B) */
+	if (rx_desc & 0x1)
+		prefetch(rxq->first_desc + rxq->next_desc_to_proc);
 	return (rxq->first_desc + rx_desc);
 }
 
-- 
1.7.9.5

